File: solutions/javascript/src/algorithm/set/Exercise6.js
import DynamicSetADT from "../../structure/dynamic/DynamicSetADT.js";

function union(set1, set2) {
  const result = new DynamicSetADT();

  // Agregamos todos los elementos de set1
  while (!set1.isEmpty()) {
    const value = set1.choose();
    result.add(value);
    set1.remove(value);
  }

  // Agregamos todos los elementos de set2
  while (!set2.isEmpty()) {
    const value = set2.choose();
    result.add(value); // `add` ya evita duplicados
    set2.remove(value);
  }

  return result;
}


function intersection(set1, set2) {
  const result = new DynamicSetADT();
  const temp = new DynamicSetADT();

  while (!set1.isEmpty()) {
    const value = set1.choose();
    if (set2.exist(value)) {
      result.add(value);
    }
    temp.add(value); // lo guardamos para no perder set1
    set1.remove(value);
  }

  // Restaurar set1 desde temp
  while (!temp.isEmpty()) {
    const value = temp.choose();
    set1.add(value);
    temp.remove(value);
  }

  return result;
}


const setA = new DynamicSetADT();
[1, 2, 3, 4].forEach(v => setA.add(v));

const setB = new DynamicSetADT();
[3, 4, 5, 6].forEach(v => setB.add(v));

const unionSet = union(setA, setB);
console.log("Unión:");
while (!unionSet.isEmpty()) {
  const value = unionSet.choose();
  console.log(value);
  unionSet.remove(value);
}

const setC = new DynamicSetADT();
[1, 2, 3, 4].forEach(v => setC.add(v));

const setD = new DynamicSetADT();
[3, 4, 5, 6].forEach(v => setD.add(v));

const intersectionSet = intersection(setC, setD);
console.log("Intersección:");
while (!intersectionSet.isEmpty()) {
  const value = intersectionSet.choose();
  console.log(value);
  intersectionSet.remove(value);
}

File: solutions/python/algorithm/set/Exercise6.py
from structure.dynamic.DynamicSetADT import DynamicSetADT


def intersection(set_a, set_b):
    result = DynamicSetADT()
    while not set_a.isEmpty():
        value = set_a.choose()
        if set_b.exist(value):
            result.add(value)
        set_a.remove(value)
    return result

def union(set_a, set_b):
    result = DynamicSetADT()
    while not set_a.isEmpty():
        value = set_a.choose()
        result.add(value)
        set_a.remove(value)

    while not set_b.isEmpty():
        value = set_b.choose()
        result.add(value)
        set_b.remove(value)

    return result

# Crear dos conjuntos
a = DynamicSetADT()
b = DynamicSetADT()

for v in [1, 2, 3]:
    a.add(v)
for v in [2, 3, 4]:
    b.add(v)

inter = intersection(a, b)
print("Intersección:")
while not inter.isEmpty():
    val = inter.choose()
    print(val, end=" ")
    inter.remove(val)

# Re-crear los conjuntos porque los anteriores fueron destruidos
a = DynamicSetADT()
b = DynamicSetADT()
for v in [1, 2, 3]:
    a.add(v)
for v in [2, 3, 4]:
    b.add(v)

un = union(a, b)
print("\nUnión:")
while not un.isEmpty():
    val = un.choose()
    print(val, end=" ")
    un.remove(val)